lib/ecs_deploy/auto_scaler/instance_fluctuation_manager.rb

```diff
@@ -0,0 +1,165 @@
+require "aws-sdk-autoscaling"
+require "aws-sdk-ec2"
+require "aws-sdk-ecs"
+
+module EcsDeploy
+  module AutoScaler
+    class InstanceFluctuationManager
+      attr_reader :logger
+
+      def initialize(region:, cluster:, cluster_to_asg:, desired_capacity:, logger: nil)
+        @region = region
+        @cluster = cluster
+        @cluster_to_asg = cluster_to_asg
+        @desired_capacity = desired_capacity
+        @logger = logger
+
+        unless cluster_to_asg.key?(cluster)
+          raise ArgumentError, "Unknown cluster: #{cluster}"
+        end
+
+        @asg_name = cluster_to_asg[cluster]
+      end
+
+      def increase
+        asg = as_client.describe_auto_scaling_groups(auto_scaling_group_names: [@asg_name]).auto_scaling_groups.first
+
+        @logger&.info("Increase desired capacity of #{@asg_name}: #{asg.desired_capacity} => #{asg.max_size}")
+        as_client.update_auto_scaling_group(auto_scaling_group_name: @asg_name, desired_capacity: asg.max_size)
+
+        # Run in background because increasing instances may take time
+        Thread.new do
+          loop do
+            cluster = ecs_client.describe_clusters(clusters: [@cluster]).clusters.first
+            instance_count = cluster.registered_container_instances_count
+            if instance_count == asg.max_size
+              @logger&.info("Succeeded to increase instance count!")
+              break
+            end
+            @logger&.info("Current registered instance count: #{instance_count}")
+            sleep 5
+          end
+        end
+      end
+
+      def decrease
+        asg = as_client.describe_auto_scaling_groups(auto_scaling_group_names: [@asg_name]).auto_scaling_groups.first
+
+        decrease_count = asg.desired_capacity - @desired_capacity
+        if decrease_count <= 0
+          @logger&.info("The capacity is already #{asg.desired_capacity}")
+          return
+        end
+        @logger&.info("Decrease desired capacity of #{@asg_name}: #{asg.desired_capacity} => #{@desired_capacity}")
+
+        container_instance_arns = ecs_client.list_container_instances(
+          cluster: @cluster
+        ).container_instance_arns
+        container_instances = ecs_client.describe_container_instances(
+          cluster: @cluster,
+          container_instances: container_instance_arns
+        ).container_instances
+
+        region_to_container_instances = container_instances.sort_by {|ci| - ci.running_tasks_count }.group_by do |ci|
+          ci.attributes.find {|attribute| attribute.name == "ecs.availability-zone" }.value
+        end
+        target_container_instances = if region_to_container_instances.empty?
```

Comment:
How about using an early return?

```suggestion
        if region_to_container_instances.empty?
          @logger&.info("There are no instances to terminate.")
          return
        end
```

Reply:
It's nice :+1: 

1517d43